File: java/src/test/java/leetcode/TwoSumTest.java
package leetcode;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import static org.junit.jupiter.api.Assertions.*;

public class TwoSumTest {
    
    private final TwoSum solution = new TwoSum();
    
    @Test
    @DisplayName("Basic case: [2,7,11,15], target=9")
    public void testBasicCase() {
        int[] nums = {2, 7, 11, 15};
        int target = 9;
        int[] result = solution.twoSum(nums, target);
        assertArrayEquals(new int[]{0, 1}, result);
    }
    
    @Test
    @DisplayName("Different positions: [3,2,4], target=6")
    public void testDifferentPositions() {
        int[] nums = {3, 2, 4};
        int target = 6;
        int[] result = solution.twoSum(nums, target);
        assertArrayEquals(new int[]{1, 2}, result);
    }
    
    @Test
    @DisplayName("Same value different indices: [3,3], target=6")
    public void testSameValueDifferentIndices() {
        int[] nums = {3, 3};
        int target = 6;
        int[] result = solution.twoSum(nums, target);
        assertArrayEquals(new int[]{0, 1}, result);
    }
    
    @Test
    @DisplayName("Negative numbers: [-1,-2,-3,-4,-5], target=-8")
    public void testNegativeNumbers() {
        int[] nums = {-1, -2, -3, -4, -5};
        int target = -8;
        int[] result = solution.twoSum(nums, target);
        assertArrayEquals(new int[]{2, 4}, result);
    }
    
    @Test
    @DisplayName("Performance: 1K elements")
    public void testPerformance1K() {
        int[] nums = new int[1000];
        for (int i = 0; i < nums.length; i++) {
            nums[i] = i;
        }
        int target = 1997;
        
        long start = System.nanoTime();
        int[] result = solution.twoSum(nums, target);
        long end = System.nanoTime();
        
        double timeMs = (end - start) / 1_000_000.0;
        System.out.printf("[1K elements] Execution time: %.3fms%n", timeMs);
        
        assertArrayEquals(new int[]{998, 999}, result);
    }
    
    @Test
    @DisplayName("Performance: 10K elements")
    public void testPerformance10K() {
        int[] nums = new int[10000];
        for (int i = 0; i < nums.length; i++) {
            nums[i] = i;
        }
        int target = 19997;
        
        long start = System.nanoTime();
        int[] result = solution.twoSum(nums, target);
        long end = System.nanoTime();
        
        double timeMs = (end - start) / 1_000_000.0;
        System.out.printf("[10K elements] Execution time: %.3fms%n", timeMs);
        
        assertArrayEquals(new int[]{9998, 9999}, result);
    }
    
    @Test
    @DisplayName("Performance: 100K elements")
    public void testPerformance100K() {
        int[] nums = new int[100000];
        for (int i = 0; i < nums.length; i++) {
            nums[i] = i;
        }
        int target = 199997;
        
        long start = System.nanoTime();
        int[] result = solution.twoSum(nums, target);
        long end = System.nanoTime();
        
        double timeMs = (end - start) / 1_000_000.0;
        System.out.printf("[100K elements] Execution time: %.3fms%n", timeMs);
        
        assertArrayEquals(new int[]{99998, 99999}, result);
    }
}

File: scala/src/test/scala/leetcode/ArraysAndHashing/TwoSumTest.scala
package leetcode.ArraysAndHashing

import org.scalatest.funsuite.AnyFunSuite

class TwoSumTest extends AnyFunSuite {
  
  test("basic case: [2,7,11,15], target=9") {
    val nums = Array(2, 7, 11, 15)
    val target = 9
    val result = TwoSum.twoSum(nums, target)
    assert(result.sameElements(Array(0, 1)))
  }
  
  test("different positions: [3,2,4], target=6") {
    val nums = Array(3, 2, 4)
    val target = 6
    val result = TwoSum.twoSum(nums, target)
    assert(result.sameElements(Array(1, 2)))
  }
  
  test("same value different indices: [3,3], target=6") {
    val nums = Array(3, 3)
    val target = 6
    val result = TwoSum.twoSum(nums, target)
    assert(result.sameElements(Array(0, 1)))
  }
  
  test("negative numbers: [-1,-2,-3,-4,-5], target=-8") {
    val nums = Array(-1, -2, -3, -4, -5)
    val target = -8
    val result = TwoSum.twoSum(nums, target)
    assert(result.sameElements(Array(2, 4)))
  }
  
  test("performance: 1K elements") {
    val nums = (0 until 1000).toArray
    val target = 1997
    
    val start = System.nanoTime()
    val result = TwoSum.twoSum(nums, target)
    val end = System.nanoTime()
    
    val timeMs = (end - start) / 1_000_000.0
    println(f"[1K elements] Execution time: $timeMs%.3fms")
    
    assert(result.sameElements(Array(998, 999)))
  }
  
  test("performance: 10K elements") {
    val nums = (0 until 10000).toArray
    val target = 19997
    
    val start = System.nanoTime()
    val result = TwoSum.twoSum(nums, target)
    val end = System.nanoTime()
    
    val timeMs = (end - start) / 1_000_000.0
    println(f"[10K elements] Execution time: $timeMs%.3fms")
    
    assert(result.sameElements(Array(9998, 9999)))
  }
  
  test("performance: 100K elements") {
    val nums = (0 until 100000).toArray
    val target = 199997
    
    val start = System.nanoTime()
    val result = TwoSum.twoSum(nums, target)
    val end = System.nanoTime()
    
    val timeMs = (end - start) / 1_000_000.0
    println(f"[100K elements] Execution time: $timeMs%.3fms")
    
    assert(result.sameElements(Array(99998, 99999)))
  }
}